projects/packages/status/src/class-status.php

```diff
@@ -261,6 +262,59 @@ public function is_staging_site() {
 		return $is_staging;
 	}
 
+	/**
+	 * If the site is in safe mode.
+	 * TODO: Fix IDC package after we move it to connection package.
+	 *
+	 * @since $$next-version$$
+	 *
+	 * @return bool
+	 */
+	public function in_safe_mode() {
+		$cached = Cache::get( 'in_safe_mode' );
+		if ( null !== $cached ) {
+			return $cached;
+		}
+		$in_safe_mode = false;
+		if ( method_exists( 'Automattic\\Jetpack\\Identity_Crisis', 'validate_sync_error_idc_option' ) && \Automattic\Jetpack\Identity_Crisis::validate_sync_error_idc_option() ) {
+			$in_safe_mode = true;
+		}
+		/**
+		 * Filters in_safe_mode check.
+		 *
+		 * @since $$next-version$$
+		 *
+		 * @param bool $in_safe_mode If the current site is in safe mode.
+		 */
+		$in_safe_mode = apply_filters( 'jetpack_is_in_safe_mode', $in_safe_mode );
+
+		Cache::set( 'in_safe_mode', $in_safe_mode );
+		return $in_safe_mode;
+	}
+
+	/**
+	 * If the site is a development/staging site.
+	 * This is a new version of is_staging_site added to separate safe and legacy staging modes.
+	 * This method only checks for core WP_ENVIRONMENT_TYPE setting
+	 *
+	 * @since $$next-version$$
+	 *
+	 * @return bool
+	 */
+	public static function is_development_site() {
+		$is_dev_site = ! in_array( wp_get_environment_type(), array( 'production', 'local' ), true );
+		/**
+		 * Filters is_development_site check.
+		 *
+		 * @since $$next-version$$
+		 *
+		 * @param bool $is_dev_site If the current site is a staging or dev site.
+		 */
+		$is_dev_site = apply_filters( 'jetpack_is_development_site', $is_dev_site );
```

Comment:
It appears that this filter is the only one that runs every time, all others methods "cache" theirs.
Should we use `Cache` class here as well?

Reply:
I deliberately avoided cache here, but I could be wrong. My reasoning was related to the PR that introduced caching: https://github.com/Automattic/jetpack/pull/22985
The `is_staging_site()` gets many calls and does many things, so caching made sense. The new `is_development_site()` is not called as often and is much simpler. Also, the downside of caching would be that changing the WP_ENVIRONMENT_TYPE value or using a filter could have a delayed effect, which doesn't sound optimal.

What do you think? If you agree, I will add comments to the method to clarify this in the code.

Reply:
This is not the actual caching, it's just a runtime storage for multiple calls within the same PHP execution. It doesn't get saved into any storage. So there shouldn't be any issues with that.

My main concern is that if the value changes somehow mid-execution (e.g. via the filter), Jetpack will get into an unexpected and untested state. We can't really have half the script run in developer mode, and then get switched to live.
It's an unlikely scenario, so I'm not sure if we should account for that. But might cause some hard to debug issues if it does happen.

Reply:
Ha, "assumption is the mother of all mistakes" rule strikes again. I assumed that Cache class stores stuff because of that previous PR that introduced it. Thanks for clarifying what it really does. I will add it then.

Reply:
Fixed in https://github.com/Automattic/jetpack/pull/37023/commits/6b7c8e2b5efb1eae175be0ea38c41a29a6800252